fdbserver/storageserver.actor.cpp

```diff
@@ -9135,382 +9135,443 @@ ACTOR Future<Void> fetchKeys(StorageServer* data, AddingShard* shard) {
 				TraceEvent(bulkLoadVerboseEventSev(), "SSBulkLoadTaskFetchKey", data->thisServerID)
 				    .detail("DataMoveId", dataMoveId.toString())
 				    .detail("Range", keys)
+				    .detail("Knobs", SERVER_KNOBS->BULK_LOAD_USE_SST_INGEST)
+				    .detail("LoadType", bulkLoadTaskState.getLoadType())
+				    .detail("SupportsSstIngestion", data->storage.getKeyValueStore()->supportsSstIngestion())
 				    .detail("Phase", "File download");
-				hold = tryGetRangeForBulkLoad(results, keys, localBulkLoadFileSets);
-				rangeEnd = keys.end;
+				state bool sstIngestionPerformed = false;
+				// If SST ingestion is enabled and we have SST files, try to ingest them directly
+				if (SERVER_KNOBS->BULK_LOAD_USE_SST_INGEST && bulkLoadTaskState.getLoadType() == BulkLoadType::SST &&
+				    data->storage.getKeyValueStore()->supportsSstIngestion()) {
+					sstIngestionPerformed = true;
+					TraceEvent(bulkLoadVerboseEventSev(), "SSBulkLoadTaskFetchKey", data->thisServerID)
+					    .detail("DataMoveId", dataMoveId.toString())
+					    .detail("Range", keys)
+					    .detail("Phase", "SST ingestion");
+					// Verify that all SST files' ranges are contained within the fetchKeys range
+					for (const auto& [range, fileSet] : *localBulkLoadFileSets) {
+						ASSERT(keys.contains(range));
+					}
+					TraceEvent("SSBulkLoadSSTIngestionStart", data->thisServerID)
+					    .detail("DataMoveId", dataMoveId.toString())
+					    .detail("FKID", fetchKeysID)
+					    .detail("Range", keys)
+					    .detail("LocalDir", bulkLoadLocalDir);
+					wait(data->storage.getKeyValueStore()->ingestSSTFiles(bulkLoadLocalDir, localBulkLoadFileSets));
+					TraceEvent("SSBulkLoadSSTIngestionComplete", data->thisServerID)
+					    .detail("DataMoveId", dataMoveId.toString())
+					    .detail("FKID", fetchKeysID)
+					    .detail("Range", keys);
+				}
+				if (!sstIngestionPerformed) {
+					hold = tryGetRangeForBulkLoad(results, keys, localBulkLoadFileSets);
+					rangeEnd = keys.end;
+				}
 			} else {
 				hold = tryGetRange(results, &tr, keys);
 				rangeEnd = keys.end;
 			}
 
 			state Key blockBegin = keys.begin;
-
```

Comment:
This part of the change is too large to reason the correctness easily. I think we can simply bypass the code section from `CODE_PROBE(true, "Fetching keys for transferred shard");` to `wait(data->storage.replaceRange(blockRange, blockData));`